gnoi/os/client.go

```diff
@@ -28,11 +32,129 @@ type Client struct {
 func NewClient(c *grpc.ClientConn) *Client {
 	return &Client{client: pb.NewOSClient(c)}
 }
 
+// Install invokes the Install RPC for the OS service.
+func (c *Client) Install(ctx context.Context, imgPath, version string, printStatus bool, validateTimeout time.Duration) error {
+	file, err := ioutil.ReadFile(imgPath)
+	if err != nil {
+		return err
+	}
+	buffer := bytes.NewBuffer(file)
+	fileSize := uint64(buffer.Len())
+	install, err := c.client.Install(ctx)
+	defer install.CloseSend()
+	if err != nil {
+		return err
+	}
+
+	if err = install.Send(&pb.InstallRequest{
+		Request: &pb.InstallRequest_TransferRequest{TransferRequest: &pb.TransferRequest{Version: version}},
+	}); err != nil {
+		return err
+	}
+	transferResp, err := install.Recv()
+	if err != nil {
+		return err
+	}
+	_, validated, err := c.validateInstallRequest(transferResp)
+	if err != nil {
+		return err
+	}
+	if validated {
+		return fmt.Errorf("OS already installed on target")
+	}
+	recvErrs := make(chan error)
+	recvValidated := make(chan bool, 1)
+	go func() {
+		validated := false
+		for !validated {
+			resp, err := install.Recv()
+			if err != nil {
+				recvErrs <- err
+				continue
+			}
+			progress, validated, err := c.validateInstallRequest(resp)
+			if err != nil {
+				recvErrs <- err
+				continue
+			}
+			if validated {
+				recvValidated <- validated
+				return
+			}
+			if printStatus {
+				fmt.Printf("%d%% transferred\n", progress/fileSize)
+			}
+		}
+	}()
+	for buffer.Len() > 0 {
+		b := make([]byte, chunkSize)
+		if len(recvErrs) > 0 {
+			return c.accumulateErrors(recvErrs)
+		}
+		_, err = buffer.Read(b)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		err = install.Send(&pb.InstallRequest{
+			Request: &pb.InstallRequest_TransferContent{TransferContent: b},
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	select {
+	case <-time.After(validateTimeout):
+		return fmt.Errorf("Validation timed out")
+	case err = <-recvErrs:
+		return fmt.Errorf("%w; %s", c.accumulateErrors(recvErrs), err.Error())
+	case <-recvValidated:
+		return nil
+	}
+}
+
+func (c *Client) validateInstallRequest(response *pb.InstallResponse) (progress uint64, validated bool, err error) {
+	switch resp := response.Response.(type) {
+	case *pb.InstallResponse_Validated:
+		validated = true
+		return
+	case *pb.InstallResponse_TransferProgress:
+		progress = resp.TransferProgress.GetBytesReceived()
+		return
+	case *pb.InstallResponse_InstallError:
+		installErr := resp.InstallError
+		if installErr.GetType() == pb.InstallError_UNSPECIFIED {
+			err = fmt.Errorf("Unspecified InstallError error: %s", installErr.GetDetail())
+			return
+		}
+		err = fmt.Errorf("InstallError occured: %s", installErr.GetType().String())
+		return
+	case *pb.InstallResponse_TransferReady:
+		return
+	default:
+		err = fmt.Errorf("Unexpected response: %T(%v)", resp, resp)
+		return
+	}
+}
+
+func (c *Client) accumulateErrors(recvErrs chan error) error {
+	err := <-recvErrs
+	if err == nil {
```

Comment:
you should be ok making this one line.

Reply:
I can but it causes golint to trigger sincei ave to put the rest in an else statement.
```go
	if err := <-recvErrs; err == nil {
		return nil
	} else {
		for len(recvErrs) > 0 {
			err = fmt.Errorf("%s; %w", (<-recvErrs).Error(), err)
		}
		return err
	}
```
`if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)go-lint`